src/analyses/dependence_graph.cpp

```diff
@@ -114,14 +112,43 @@ void dep_graph_domaint::control_dependencies(
         post_dom_all=false;
     }
 
-    if(post_dom_all ||
-       !post_dom_one)
-      control_deps.erase(it);
-
-    it=next;
+    if(post_dom_all || !post_dom_one)
+    {
+      control_deps.erase(cd);
+    }
+    else
+    {
+      control_deps.insert(cd);
+    }
   }
 }
 
+bool dep_graph_domaint::merge_control_dependencies(
+  const depst &other_control_deps,
+  const depst &other_control_dep_candidates)
+{
+  bool changed = false;
+
+  // Merge control dependencies
+
+  size_t n = control_deps.size();
+
+  control_deps.insert(other_control_deps.begin(), other_control_deps.end());
```

Comment:
It seems a shame to degrade the previous more-efficient algorithm that makes a single walk over the target set into this one that makes |other_control_deps| searches from the root. If it bothers you that the old algorithm is less clear, how about factoring it into a util called `union_ordered_range` or something?

Reply:
Indeed. I assumed that it would implement the same algorithm if both ranges are sorted but it's not the case. I'll add a `union_ordered_range` utility function.